File: GAIL_Python/Algorithms/IntegrationExpectation/meanYParam.py
import inspect
import numbers
from operator import attrgetter

import numpy as np
try:
    from GAIL_Python.Algorithms.IntegrationExpectation.CubMeanParam import CubMeanParam
except:
    from CubMeanParam import CubMeanParam
try:
    from GAIL_Python.Algorithms.IntegrationExpectation.ErrorParam import ErrorParam
except:
    from ErrorParam import ErrorParam


def default_random_generator(n):
    """
    X_i^2 where X_i comes from a U[0,1]
    :param n: n is length of random numbers needed
    :return: vector of random numbers
    """
    return np.random.uniform(size=n) ** 2


class MeanYParam(object):
    """
    MeanYParam is a class containing the parameters related to
    algorithms that find the mean of a random variable
    This class contains the random number generator, the uncertainty
    """

    @property
    def Yout(self):
        if self._Yout is None:
            self._Yout = len(self.y(1))
        return self._Yout

    @Yout.setter
    def Yout(self, Yout_value):
        self._Yout = Yout_value

    Y = property(attrgetter('_Y'))

    @Y.setter
    def Y(self, Y_func):
        # ISFCN To judge if input is a function handle or not
        if not callable(Y_func):
            raise Exception("y has to be a function")
        self._Y = Y_func

    alpha = property(attrgetter('_alpha'))

    @alpha.setter
    def alpha(self, alpha_value):  # {'numeric'}, {'scalar','nonnegative', '<', 1})
        msg = "alpha should be positive and less than 1"
        if alpha_value is None:
            raise Exception("alpha is empty. " + msg)
        if not isinstance(alpha_value, numbers.Number):
            raise Exception("alpha value should be a number. " + msg)
        else:
            if alpha_value < 0:
                raise Exception("alpha is negative. " + msg)
            elif alpha_value > 1:
                raise Exception("alpha is greater than 1. " + msg)
        self._alpha = alpha_value

    nSig = property(attrgetter('_nSig'))

    @nSig.setter
    def nSig(self, nSig_value):  # {'numeric'}, {'scalar','nonnegative'})
        msg = "nSig should be a positive number"
        if nSig_value is None:
            raise Exception("nSig is empty. " + msg)
        if not isinstance(nSig_value, numbers.Number):
            raise Exception("nSig is not a number. " + msg)
        else:
            if nSig_value < 0:
                raise Exception("nSig is negative. " + msg)
        self._nSig = nSig_value

    absTol = property(attrgetter('_absTol'))

    @absTol.setter
    def absTol(self, absTol_value):  # {'numeric'}, {'scalar','nonnegative', '<', 1})
        msg = "absTol should be positive number"
        if absTol_value is None:
            raise Exception("value cannot be empty. " + msg)
        if not isinstance(absTol_value, numbers.Number):
            raise Exception("absTol value should be a number. " + msg)
        else:
            if absTol_value < 0:
                raise Exception("absTol value cannot be negative. " + msg)
        self._absTol = absTol_value

    relTol = property(attrgetter('_relTol'))

    @relTol.setter
    def relTol(self, relTol_value):  # {'numeric'}, {'scalar','nonnegative'})
        msg = "relTol should be a positive number"
        if relTol_value is None:
            raise Exception("value cannot be empty. " + msg)
        if not isinstance(relTol_value, numbers.Number):
            raise Exception("relTol value should be a number. " + msg)
        else:
            if relTol_value < 0:
                raise Exception("relTol value cannot be negative. " + msg)
        self._relTol = relTol_value

    def __init__(self, Y=None, absTol=1e-2, relTol=0, alpha=0.01, nSig=1000, **kwargs):
        if isinstance(Y, MeanYParam):
            # copy the MeanYParam object
            self.make_copy(Y)
        else:
            if Y is None:
                self.Y = default_random_generator
            else:
                self.Y = Y  # random number generator

            self.absTol = absTol
            self.relTol = relTol

            self.alpha = alpha  # uncertainty
            self.nSig = nSig  # sample size to estimate variance

            self.err = ErrorParam(absTol=absTol, relTol=relTol, **kwargs)  # an errorParam object
            self.CM = CubMeanParam(**kwargs)  # a cubMeanParam object
            try:
                self.nY = kwargs['nY']
            except:
                self.nY = 1  # number of Y for each mean

    def make_copy(self, Y):
        self.Y = Y.Y
        self.absTol = Y.absTol
        self.relTol = Y.relTol
        self.alpha = Y.alpha  # uncertainty
        self.nSig = Y.nSig  # sample size to estimate variance
        self.err = ErrorParam(absTol=Y.err.absTol, relTol=Y.err.relTol)
        self.CM = CubMeanParam(inflate=Y.CM.inflate, inflateFun=Y.CM.inflateFun, nInit=Y.CM.nInit, nMax=Y.CM.nMax,
                               nMu=Y.CM.nMu, trueMuCV=Y.CM.trueMuCV)
        self.nY = Y.nY

    def __str__(self):

        try:
            function_source = inspect.getsource(self.Y)
        except:
            function_source = str(self.Y)

        return 'meanYParam with properties:\n\tY\t: {}\n\tabsTol\t: {}\n\trelTol\t: {}\n\talpha\t: {}\n\tnSig\t: {}' \
            .format(function_source, self.absTol, self.relTol, self.alpha, self.nSig)

File: Algorithms/+gail/meanYParam.m
classdef meanYParam < handle & matlab.mixin.CustomDisplay
   %GAIL.MEANYPARAM is a class containing the parameters related to
   %algorithms that find the mean of a random variable
   %   This class contains the random number generator, the uncertainty
   %
   % Example 1. Construct a meanYParam object with default parameters
   % >> meanYParamObj = gail.meanYParam
   % meanYParamObj = 
   %   meanYParam with properties:
   % 
   %            Y: @(n)rand(n,1)
   %       absTol: 0.0100
   %       relTol: 0
   %        alpha: 0.0100
   %
   %
   % Example 2. Construct a cubParam object with properly ordered inputs
   % >> meanYParamObj = gail.meanYParam(@(n) sum(rand(n,4).^3,2),0.001)
   % meanYParamObj = 
   %   meanYParam with properties:
   % 
   %            Y: @(n)sum(rand(n,4).^3,2)
   %       absTol: 1.0000e-03
   %       relTol: 0
   %        alpha: 0.0100
   %
   %
   % Example 3. Using name/value pairs
   % >> meanYParamObj = gail.meanYParam('nSig', 1e4, 'Y', @(n)sin(sum(rand(n,4).^3,2)), 'relTol', 0.1)
   % meanYParamObj = 
   %   meanYParam with properties:
   % 
   %            Y: @(n)sin(sum(rand(n,4).^3,2))
   %       absTol: 0.0100
   %       relTol: 0.1000
   %        alpha: 0.0100
   %         nSig: 10000
   %
   %
   % Example 4. Using a structure for input
   % >> inpStruct.Y = @(n) sin(sum(rand(n,2),2));
   % >> inpStruct.nSig = 1e4;
   % >> inpStruct.relTol = 0.1;
   % >> meanYParamObj = gail.meanYParam(inpStruct)
   % meanYParamObj = 
   %   meanYParam with properties:
   % 
   %            Y: @(n)sin(sum(rand(n,2),2))
   %       absTol: 0.0100
   %       relTol: 0.1000
   %        alpha: 0.0100
   %         nSig: 10000
   %
   %
   % Example 5. Copying a meanYParam object and changing some properties
   % >> NewMeanYParamObj = gail.meanYParam(meanYParamObj,'Y',@(n) rand(n,3))
   % NewMeanYParamObj = 
   %   meanYParam with properties:
   % 
   %            Y: @(n)rand(n,3)
   %       absTol: 0.0100
   %       relTol: 0.1000
   %        alpha: 0.0100
   %         nSig: 10000
   %
   %
   % Author: Fred J. Hickernell

   properties
      Y %random number generator
      alpha %uncertainty
      err %an errorParam object
      CM %a cubMeanParam object
      nSig %sample size to estimate variance
      nY %number of Y for each mean
   end
   
    properties (Dependent = true)
       nYOut %number of Y outputs
    end
   
   properties (Hidden, SetAccess = private)
      def_Y = @(n) rand(n,1) %default random number generator
      def_alpha = 0.01 %default uncertainty
      def_nSig = 1000 %default uncertainty
      def_nY = 1 %default number of Y per mean
   end
   
   
   methods
      
      % Creating a meanYParam process
      function obj = meanYParam(varargin)
         %this constructor essentially parses inputs
         %the parser will look for the following in order
         %  # a copy of a meanYParam object
         %  # a function
         %  # a structure
         %  # numbers: absTol, relTol
         %  # name-value pairs
         
         start = 1;
         useDefaults = true;
         objInp = 0;
         YInp = 0;
         structInp = 0;
         if nargin %there are inputs to parse and assign
            if isa(varargin{start},'gail.meanYParam') 
               %the first input is a meanYParam object so copy it
               objInp = start;
               start = start + 1;
            end
            if nargin >= start
               if isstruct(varargin{start}) %next input is a structure containing Y
                  structInp = start;
                  start = start + 1;
               end
            end
            if nargin >= start
               if gail.isfcn(varargin{start}) %next input is the function Y
                  YInp = start;
                  start = start + 1;
               end
            end
         end
         
         if objInp
            val = varargin{objInp}; %first input
            obj.err = gail.errorParam(val.err); %copy errorParam object
            obj.CM = gail.cubMeanParam(val.CM); %copy cubMeanParam object
            obj.Y = val.Y; %copy random number generator
            obj.alpha = val.alpha; %copy uncertainty
            obj.nSig = val.nSig; %copy sample size for sigma
            obj.nY = val.nY; %copy number of Y values per mu
            useDefaults = false;
         end

         %Parse errorParam properties
         whichErrParse = [structInp start:nargin];
         whichErrParse = whichErrParse(whichErrParse > 0);
         if objInp
            obj.err = gail.errorParam(obj.err,varargin{whichErrParse});
         else
            obj.err = gail.errorParam(varargin{whichErrParse});
         end

         %Parse cubMeanParam properties
         whichCMParse = [structInp start:nargin];
         whichCMParse = whichCMParse(whichCMParse > 0);
         if objInp
            obj.CM = gail.cubMeanParam(obj.CM,varargin{whichCMParse});
         else
            obj.CM = gail.cubMeanParam(varargin{whichCMParse});
         end

         %Now begin to parse inputs
         p = inputParser; %construct an inputParser object
         p.KeepUnmatched = true; %ignore those that do not match
         p.PartialMatching = false; %don'try a partial match
         p.StructExpand = true; %expand structures
         done = false; %not finished parsing
         if nargin >= start
            if ischar(varargin{start})
               %there may be input string/value pairs or a structure
               MATLABVERSION = gail.matlab_version;
               if MATLABVERSION >= 8.3
                  f_addParamVal = @addParameter;
               else
                  f_addParamVal = @addParamValue;
               end
               parseRange = start:nargin;
               done = true;
            end
         end
         if ~done %then nothingleft or just numbers
           f_addParamVal = @addOptional;
           parseRange = []; %to account for the two tolerances already parsed
         end
         f_addParamVal(p,'Y',obj.def_Y);
         f_addParamVal(p,'alpha',obj.def_alpha);
         f_addParamVal(p,'nSig',obj.def_nSig);
         f_addParamVal(p,'nY',obj.def_nY);
         
         if structInp
            parse(p,varargin{parseRange},varargin{structInp}) 
            %parse inputs with a structure
         else
            parse(p,varargin{parseRange}) %parse inputs w/o structure
         end
         struct_val = p.Results; %store parse inputs as a structure
         if ~useDefaults %remove defaults if copying cubParam object
            struct_val = rmfield(struct_val,p.UsingDefaults);
         end
         
         %Assign values of structure to corresponding class properties
         if YInp
            obj.Y = varargin{YInp}; %assign function
         elseif isfield(struct_val,'Y')
            obj.Y = struct_val.Y;
         end
         
         if isfield(struct_val,'alpha')
            obj.alpha = struct_val.alpha;
         end
         if isfield(struct_val,'nSig')
            obj.nSig = struct_val.nSig;
         end
         if isfield(struct_val,'nY')
            obj.nY = struct_val.nY;
         end
         
      end %of constructor
     
      function set.Y(obj,val)
         validateattributes(val, {'function_handle'}, {})
         obj.Y = val;
      end
      
      function set.alpha(obj,val)
         validateattributes(val, {'numeric'}, {'scalar','nonnegative', ...
            '<', 1})
         obj.alpha = val;
      end
                       
      function set.nSig(obj,val)
         validateattributes(val, {'numeric'}, {'scalar','positive','integer'})
         obj.nSig = val;
      end
                                                    
      function set.nY(obj,val)
         validateattributes(val, {'numeric'}, {'positive','integer'})
         obj.nY = val;
      end
                                              
      function val = get.nYOut(obj)
         val = numel(obj.Y(1)); 
      end         
     
      
   end
   
   methods (Access = protected)
   
      function outval = checkNY(obj,inval)
         assert(obj.fun.nYOut - sum(inval) == obj.CM.nCV)
         outval = inval;
      end
    
      function propgrp = getPropertyGroups(obj)
        if ~isscalar(obj)
           propgrp = getPropertyGroups@matlab.mixin.CustomDisplay(obj);
        else
           propList = getPropertyList(obj);
           propgrp = matlab.mixin.util.PropertyGroup(propList);
        end
     end
      
     function propList = getPropertyList(obj)
         propList = struct('Y', obj.Y, ...
            'absTol', obj.err.absTol, ...
            'relTol', obj.err.relTol, ...
            'alpha', obj.alpha);
         if obj.nSig ~= obj.def_nSig
            propList.nSig = obj.nSig;
         end
         if obj.CM.nMax ~= obj.CM.def_nMax
            propList.nMax = obj.CM.nMax;
         end
         if obj.CM.nMu ~= obj.CM.def_nMu
            propList.nMu = obj.CM.nMu;
         end
         if obj.nY ~= obj.def_nY
            propList.nY = obj.nY;
         end
         if numel(obj.CM.trueMuCV)
            propList.trueMuCV = obj.CM.trueMuCV;
         end

      end
      
   end
 
   
end

